### DISCUSSION TITLE
0.23.0/detection/tools/line_zone/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#1548] [Docs]
# 0.23.0/detection/tools/line_zone/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/0.23.0/detection/tools/line_zone/

**Comment by deepaks11** [comment]
I don’t know what I’m missing here. The intersection is not working. Any guesses on what mistake I’m making:                

```python
import supervision as sv
from supervision.geometry.core import Position
import os


os.environ["KMP_DUPLICATE_LIB_OK"]="TRUE"


class YoloObjectDetection:

    def __init__(self, model):

        self.model = model
        self.frame = None
        self.detections = None
        self. color_codes = ["#FF0000"]  # Red

        self.start, self.end = sv.Point(x=0, y=1080), sv.Point(x=2600, y=1080)
        self.line_zone = sv.LineZone(start=self.start, end=self.end)

    def predict(self,  q_img):
        try:
            frame = q_img.get()

            box_annotator = sv.BoxAnnotator()
            line_annotator = sv.LineZoneAnnotator()

            label_annotator = sv.LabelAnnotator(text_padding=8, text_position=Position.CENTER, text_scale=1, text_thickness=2)

            result = self.model.track(source=frame, agnostic_nms=True, iou=0.5, imgsz=640, conf=0.75, persist=True, verbose=False, classes=0)

            result = result[0]

            detections = sv.Detections.from_ultralytics(result)

            if detections:

                if result.boxes.id is not None:
                    detections.tracker_id = result.boxes.id.cpu().numpy().astype(int)

                labels = [
                    f"{tracker_id} {self.model.names[class_id]}"
                    for box, mask, confidence, class_id, tracker_id, class_name
                    in detections
                ]
                self.line_zone.trigger(detections)
                frame = line_annotator.annotate(frame=frame, line_counter=self.line_zone)
                frame = box_annotator.annotate(
                    scene=frame,
                    detections=detections
                )

                frame = label_annotator.annotate(
                    scene=frame, detections=detections, labels=labels)

                return frame
            else:
                return frame
        except Exception as er:
            print(er)
```

**Reply by LinasKo** [reply]
Hi @deepaks11 👋 

Did you try tracking it with our ByteTrack? I don't know much about Ultralytics. I don't know if the results from `model.track()` are similar to `model()`.

Here's how you'd do it with our tracker:
https://supervision.roboflow.com/latest/how_to/track_objects/

Next, Detections will always return an object so you should use `if detections.is_empty()` instead of `if detections`.

Then, I see the line zone is defined at y=1080. Is that the bottom of the screen? If so, I'd move the line higher up. LineZone expect the detdction to be detected on both sides of the line - it measures how many **crossed** it.

Finally, I would not expect any exceptions in this process. If there are any - you're doing something wrong.

Hope this helps!

**Comment by HugoSun** [comment]
If I want only specific detections to trigger an event (e.g., only certain categories trigger the event while others do not), how should I write the code to implement this condition?

**Reply by LinasKo** [reply]
In this case, it'd be a good idea to filter the detections first.

```python
filtered_detections = detections[detections.class_id == 0]
line_zone.trigger(filtered_detections)
```

**Comment by 7ser23** [comment]
hi, I am having some issues understanding the trigger method.

when i do:
```print(zone_line.in_count, zone_line.out_count)``` 
I get the detections that crossed my line. But when I also try to see the output of trigger: ```print(zone_line.trigger(detections=detections[detections.class_id == 0]))```
I get an array (size depending on the number of detections) as always false even when the counter has more than one.

the code is like this:
``` python

zone_line = sv.LineZone(start=LINE_START, end=LINE_END)

# Create line annotator
zone_line_annotator = sv.LineZoneAnnotator(thickness=1, text_thickness=2, text_scale=1, color=sv.Color.WHITE)

    for frame in frames_generator:
        results = model.infer(frame, confidence=confidence, iou_threshold=iou)[0]
        detections = sv.Detections.from_inference(results)
        #detections = detections[find_in_list(detections.class_id, classes)]
        detections = tracker.update_with_detections(detections)

        annotated_frame = frame.copy()

        annotated_frame = box_annotator.annotate(
            scene=annotated_frame, detections=detections[detections.class_id == 0])

        labels = [f"#{tracker_id} {confidence:0.2f}"
                  for confidence,tracker_id
                  in zip(detections.confidence, detections.tracker_id)
        ]
        annotated_frame = label_annotator.annotate(
            scene=annotated_frame, detections=detections, labels=labels)
        ## polygon that detects leaving customers
        #zone_out.trigger(detections=detections)

        ## Line that detects leaving customers
        zone_line.trigger(detections=detections[detections.class_id == 0])


        # update line counter
        annotated_frame = zone_line_annotator.annotate(annotated_frame, line_counter=zone_line)
        print(zone_line.trigger(detections=detections[detections.class_id == 0]))
        #print(detections[detections.class_id == 0].tracker_id)
        print(zone_line.in_count, zone_line.out_count)
```

I think that I should be expecting a True when they cross the line right?

**Reply by LinasKo** [reply]
Hi @7ser23 👋 

Because you are triggering twice, the line zone thinks the detections haven't moved, and therefore return `False`. Store the value of the first trigger (in `## Line that detects leaving customers` section) and print it to see the actual trigger value. Then, remove the second call to `trigger` in `# update line counter` section.

This should fix the issue you're having 😉

**Reply by 7ser23** [reply]
U awesome, it works. Thank you very much. Now I can identify per polygon and line which detection left my queue.
